File: 백준/Gold/1238. 파티/파티.py
#3:35
import sys
from heapq import heappop, heappush

N,M,X = map(int,sys.stdin.readline().split())

dic = [[] for _ in range(N)]
rev_dic = [[] for _ in range(N)]

for _ in range(M):
  a,b,d = map(int,sys.stdin.readline().split())
  a,b = a-1, b-1
  dic[a].append([b,d])
  rev_dic[b].append([a,d])
    
def dijkstra(s, dic):
  q = []
  result = [float('INF')]*N
  heappush(q, (0,s))
  result[s] = 0
  while(q):
    cd, cx = heappop(q)
    for nx, nd in dic[cx]:
      if result[nx] > cd + nd:
        result[nx] = cd+nd
        heappush(q,(cd+nd,nx)) 
  return result


back = dijkstra(X-1, dic)
go = dijkstra(X-1, rev_dic)
answer = 0
for i in range(len(go)):
  answer = max(answer,go[i]+back[i])

print(answer)

    
    
    
    

    

File: 백준/Gold/1238. 파티/파티.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin','utf8').trim().split('\n')
const [N,M,X] = input.shift().trim().split(' ').map(Number)

const graph = [...Array(N+1)].map(()=>[])
const reverse_graph = [...Array(N+1)].map(()=>[])

for (let data of input){
    let [x,y,t]=data.trim().split(' ').map(Number)
    graph[x].push([y,t])
    reverse_graph[y].push([x,t])
}

function bfs(x,graph){
    let dp = new Array(N+1).fill(Infinity)
    let q = [x]
    dp[q] = 0

    while(q.length){
        let cx = q.shift()
        for (let [nx,nt] of graph[cx]){
            if (dp[nx] > dp[cx]+nt){
                dp[nx] = dp[cx]+nt
                q.push(nx)
            }
        }
    }
    
    return dp
}

let go = bfs(X,reverse_graph)
let back = bfs(X, graph)

let answer = 0
for (let i=1;i<=N;i++){
    answer = Math.max(go[i]+back[i],answer)
}

console.log(answer)